data-prepper-plugins/in-memory-source-coordination-store/src/main/java/org/opensearch/dataprepper/plugins/sourcecoordinator/inmemory/InMemorySourceCoordinationStore.java

```diff
@@ -50,41 +50,55 @@ public void initializeStore() {
     }
 
     @Override
-    public Optional<SourcePartitionStoreItem> getSourcePartitionItem(final String partitionKey) {
-        return inMemoryPartitionAccessor.getItem(partitionKey);
+    public Optional<SourcePartitionStoreItem> getSourcePartitionItem(final String sourceIdentifier, final String partitionKey) {
+        synchronized (this) {
+            return inMemoryPartitionAccessor.getItem(sourceIdentifier, partitionKey);
+        }
     }
 
     @Override
-    public boolean tryCreatePartitionItem(final String partitionKey, final SourcePartitionStatus sourcePartitionStatus, final Long closedCount, final String partitionProgressState) {
-
-        if (inMemoryPartitionAccessor.getItem(partitionKey).isEmpty()) {
-            final InMemorySourcePartitionStoreItem inMemorySourcePartitionStoreItem = new InMemorySourcePartitionStoreItem();
-            inMemorySourcePartitionStoreItem.setSourcePartitionKey(partitionKey);
-            inMemorySourcePartitionStoreItem.setSourcePartitionStatus(sourcePartitionStatus);
-            inMemorySourcePartitionStoreItem.setClosedCount(closedCount);
-            inMemorySourcePartitionStoreItem.setPartitionProgressState(partitionProgressState);
-            inMemoryPartitionAccessor.queuePartition(inMemorySourcePartitionStoreItem);
-            return true;
+    public boolean tryCreatePartitionItem(final String sourceIdentifier,
+                                          final String partitionKey,
+                                          final SourcePartitionStatus sourcePartitionStatus,
+                                          final Long closedCount,
+                                          final String partitionProgressState) {
+        synchronized (this) {
```

Comment:
Each sub-pipeline (sourceIdentifier in this case) will get it's own copy of the SourceCoordinator and SourceCoordinationStore. So, with the current sources that only run a single thread (s3 scan), there will be no synchronization at all even with multiple sub-pipelines and sources, since they each have their own copy for the store. Adding the synchronized blocks ensures that multiple threads could be run in the source to process partitions. I'm sure we could optimize/tighten this synchronization a bit, but for simplicity I'm proposing we leave it as is here until we have a use case